src/main/java/org/truffleruby/debug/TruffleDebugNodes.java

```diff
@@ -178,44 +202,105 @@ private DynamicObject ast(Node node) {
     public abstract static class ASTGraphNode extends CoreMethodArrayArgumentsNode {
 
+        @TruffleBoundary
         @Specialization(guards = "isRubyMethod(method)")
-        public DynamicObject astMethod(DynamicObject method) {
-            return ast(Layouts.METHOD.getMethod(method));
+        public DynamicObject astMethod(DynamicObject method, NotProvided block) {
+            astGraph(Layouts.METHOD.getMethod(method));
+            return nil();
         }
 
+        @TruffleBoundary
         @Specialization(guards = "isRubyUnboundMethod(method)")
-        public DynamicObject astUnboundMethod(DynamicObject method) {
-            return ast(Layouts.UNBOUND_METHOD.getMethod(method));
+        public DynamicObject astUnboundMethod(DynamicObject method, NotProvided block) {
+            astGraph(Layouts.UNBOUND_METHOD.getMethod(method));
+            return nil();
         }
 
+        @TruffleBoundary
         @Specialization(guards = "isRubyProc(proc)")
-        public DynamicObject astProc(DynamicObject proc) {
-            return ast(Layouts.PROC.getMethod(proc));
+        public DynamicObject astProc(DynamicObject proc, NotProvided block) {
+            astGraph(Layouts.PROC.getSharedMethodInfo(proc).getName(), Layouts.PROC.getCallTargetForType(proc));
+            return nil();
         }
 
         @TruffleBoundary
-        private DynamicObject ast(InternalMethod method) {
-            if (method.getCallTarget() instanceof RootCallTarget) {
-                return ast(method.getName(), ((RootCallTarget) method.getCallTarget()).getRootNode());
+        @Specialization
+        public DynamicObject astBlock(NotProvided proc, DynamicObject block) {
+            astGraph(Layouts.PROC.getSharedMethodInfo(block).getName(), Layouts.PROC.getCallTargetForType(block));
+            return nil();
+        }
+
+        private void astGraph(InternalMethod method) {
+            astGraph(method.getName(), (RootCallTarget) method.getCallTarget());
+        }
+
+        private void astGraph(String name, CallTarget callTarget) {
+            if (callTarget instanceof RootCallTarget) {
+                astGraph(name, (RootCallTarget) callTarget);
             } else {
-                return nil();
+                throw new RaiseException(getContext().getCoreExceptions().internalError("call target is not a root call target", this));
             }
         }
 
-        private DynamicObject ast(String name, Node node) {
-            if (node != null) {
-                GraphPrintVisitor graphPrinter = new GraphPrintVisitor();
-                graphPrinter.beginGraph(name).visit(node);
+        private void astGraph(String name, RootCallTarget callTarget) {
+            final GraphPrintVisitor graphPrinter = new GraphPrintVisitor();
+            graphPrinter.beginGraph(name).visit(callTarget.getRootNode());
+            graphPrinter.printToNetwork(true);
+            graphPrinter.close();
```

Comment:
GraphPrintVisitor is deprecated in latest Truffle, we'll have to take care when updating Truffle:
https://github.com/graalvm/graal/blob/8bb574448ec64548e6f8796743c5385fc3d7d782/truffle/src/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/GraphPrintVisitor.java